--- MQTTSNGateway/src/MQTTSNGWClientRecvTask.cpp.orig
+++ MQTTSNGateway/src/MQTTSNGWClientRecvTask.cpp
@@ -149,7 +149,9 @@ void ClientRecvTask::run()
         {
             log(client, packet, 0);
 
-            if (client->isDisconnect() && packet->getType() != MQTTSN_CONNECT)
+            if ((client->isDisconnect() && packet->getType() != MQTTSN_CONNECT) ||
+                (client->isConnecting() && packet->getType() != MQTTSN_WILLTOPIC &&
+                 packet->getType() != MQTTSN_WILLMSG))
             {
                 WRITELOG("%s MQTTSNGWClientRecvTask %s is not connecting.%s\n",
                 ERRMSG_HEADER, client->getClientId(), ERRMSG_FOOTER);
