--- MQTTSNGateway/src/linux/xbee/SensorNetwork.cpp.orig
+++ MQTTSNGateway/src/linux/xbee/SensorNetwork.cpp
@@ -214,11 +214,11 @@
 int XBee::broadcast(const uint8_t* payload, uint16_t payloadLen){
 	SensorNetAddress addr;
 	addr.setBroadcastAddress();
-	return send(payload, (uint8_t) payloadLen, &addr);
+	return send(payload, payloadLen, &addr);
 }
 
 int XBee:: unicast(const uint8_t* payload, uint16_t payloadLen, SensorNetAddress* addr){
-	return send(payload, (uint8_t) payloadLen, addr);
+	return send(payload, payloadLen, addr);
 }
 
 int XBee::recv(uint8_t* buf, uint16_t bufLen, SensorNetAddress* clientAddr)
@@ -270,7 +270,7 @@
 	uint8_t buf;
 	uint8_t pos = 0;
 	uint8_t checksum = 0;
-	uint8_t len = 0;
+	uint16_t len = 0;
 
 	 while ( _serialPort->recv(&buf) )
 	 {
@@ -291,6 +291,10 @@
     {
     	goto errexit;
     }
+    else
+    {
+        len = (uint16_t)buf << 8;
+    }

     if ( recv(&buf) < 0 ) // LSB length
 	{
@@ -298,9 +301,15 @@
 	}
 	else
 	{
-		len = buf;
+		len += buf;
 	}
 
+	/* We only have a 256-byte buffer in recv(), so ensure we do not overflow it. */
+	if (len > 255)
+	{
+		goto errexit;
+	}
+
     pos = 0;
     while ( len-- )
     {
@@ -327,14 +336,17 @@
 	return -1;
 }

-int XBee::send(const uint8_t* payload, uint8_t pLen, SensorNetAddress* addr){
+int XBee::send(const uint8_t* payload, uint16_t pLen, SensorNetAddress* addr){
 	D_NWSTACK("\r\n===> Send:    ");
     uint8_t checksum = 0;
     _respCd = -1;

     _serialPort->send(START_BYTE);
-    send(0x00);              // Message Length
-    send(14 + pLen);         // Message Length
+    uint16_t frameDataLen = 14 + pLen;
+    if (frameDataLen > 255)
+        return -1;
+    send((uint8_t)(frameDataLen >> 8));    // Message Length MSB
+    send((uint8_t)(frameDataLen & 0xFF));  // Message Length LSB
 
     _serialPort->send(API_XMITREQUEST); // Transmit Request API
     checksum += API_XMITREQUEST;
--- MQTTSNGateway/src/linux/xbee/SensorNetwork.h.orig
+++ MQTTSNGateway/src/linux/xbee/SensorNetwork.h
@@ -94,7 +94,7 @@
 private:
 	int readApiFrame(uint8_t* recvData);
 	int recv(uint8_t* buf);
-	int send(const uint8_t* payload, uint8_t pLen, SensorNetAddress* addr);
+	int send(const uint8_t* payload, uint16_t pLen, SensorNetAddress* addr);
 	void send(uint8_t b);
 
 	Semaphore _sem;
